Refactor Metadata to remove Configuration global variable.

akramtexas opened the pull request:
Fix #2023

akramtexas commented:
Rebased.

akramtexas commented:
Rebased.

michaelestrin commented:
recheck

akramtexas commented:
Rebased.

akramtexas commented:
Rebased.

michaelestrin commented on cmd/core-metadata/main.go:
```diff
@@ -19,7 +19,10 @@ import (
 	"github.com/edgexfoundry/edgex-go"
 	"github.com/edgexfoundry/edgex-go/internal"
 	"github.com/edgexfoundry/edgex-go/internal/core/metadata"
+	metadataConfig "github.com/edgexfoundry/edgex-go/internal/core/metadata/config"
```
Nitpick.  Should not be aliased.

michaelestrin commented on cmd/core-metadata/toml_test.go:
```diff
@@ -8,14 +8,14 @@
 package main
 
 import (
+	metadataConfig "github.com/edgexfoundry/edgex-go/internal/core/metadata/config"
```
Nitpick.  Imports not sorted.

michaelestrin commented on internal/core/metadata/init.go:
```diff
@@ -18,8 +18,9 @@ import (
 	"sync"
 
+	"github.com/edgexfoundry/edgex-go/internal/core/metadata/container"
 	metadataContainer "github.com/edgexfoundry/edgex-go/internal/core/metadata/container"
```
You have two different references to the same package; one aliased, one not.  Recommend removing the aliased one (`metadataContainer`) as doing so would be consistent with similar refactorings in other core services.

michaelestrin commented on internal/core/metadata/rest_addressable.go:
```diff
@@ -35,9 +36,10 @@ func restGetAllAddressables(
+	config config.ConfigurationStruct) {
 
-	op := addressable.NewAddressableLoadAll(Configuration.Service, dbClient, loggingClient)
+	op := addressable.NewAddressableLoadAll(config.Service, dbClient, loggingClient)
```
To be consistent with other refactorings, use `configuration` instead of `config` for those variables that you're passing down the call chain to replace the now-removed `Configuration` global.

Mentioned here; occurs elsewhere.

michaelestrin commented on internal/core/metadata/rest_addressable_test.go:
```diff
@@ -55,8 +56,9 @@ var ErrorPathParam = "%zz"
-	Configuration = &ConfigurationStruct{Service: config.ServiceInfo{MaxResultCount: 10}}
-	defer func() { Configuration = &ConfigurationStruct{} }()
+	configuration := metadataConfig.ConfigurationStruct{
+		Service: config.ServiceInfo{MaxResultCount: 10}}
```
Clean this up:

```
	configuration := metadataConfig.ConfigurationStruct{
		Service: config.ServiceInfo{MaxResultCount: 10},
	}
```

Commented here; occurs elsewhere.

michaelestrin commented on internal/core/metadata/rest_addressable_test.go:
```diff
@@ -55,8 +56,9 @@ var ErrorPathParam = "%zz"
-	defer func() { Configuration = &ConfigurationStruct{} }()
+	configuration := metadataConfig.ConfigurationStruct{
+		Service: config.ServiceInfo{MaxResultCount: 10}}
+	defer func() { configuration = metadataConfig.ConfigurationStruct{} }()
```
Why?

michaelestrin commented on internal/core/metadata/rest_addressable_test.go:
```diff
@@ -91,14 +93,14 @@ func TestGetAllAddressables(t *testing.T) {
 				tt.dbMock,
-				errorconcept.NewErrorHandler(loggerMock))
+				errorconcept.NewErrorHandler(loggerMock),
+				configuration)
```
Move your definition of `configuration` into the test loop so each iteration has its own cleanly initialized instance.

michaelestrin commented on internal/core/metadata/rest_device.go:
```diff
@@ -707,7 +735,17 @@ func restSetDeviceLastConnectedByName(
+		dbClient,
+		errorHandler,
+		nc,
+		config); err != nil {
```
This all fits within the 120 character limit.

```
	if err = setLastConnected(d, lc, false, w, ctx, loggingClient, dbClient, errorHandler, nc, config); err != nil {
		return
	}
```
If for some reason it didn't, my preferred way to preserve readability would be to break the assignment out of the if; i.e. 

```
	err = setLastConnected(d, lc, false, w, ctx, loggingClient, dbClient, errorHandler, nc, config);
	if err != nil {
		return
	}
```

or 

```
	err = setLastConnected(
		d,
		lc,
		false,
		w,
		ctx,
		loggingClient,
		dbClient,
		errorHandler,
		nc,
		config);
	if err != nil {
		return
	}
```
Commented once; occurs elsewhere.

akramtexas replied:
Did not fit, hence the use of the second option suggested above 👍

michaelestrin commented on internal/core/metadata/rest_addressable.go:
```diff
@@ -35,9 +36,10 @@ func restGetAllAddressables(
 	dbClient interfaces.DBClient,
-	errorHandler errorconcept.ErrorHandler) {
+	errorHandler errorconcept.ErrorHandler,
+	config config.ConfigurationStruct) {
```
Pass `config.ConfigurationStruct` down the call stack by reference, not by value.

Commented here; occurs elsewhere.

akramtexas replied:
- My understanding was that interfaces are passed as references (_by default_). By looking up the online wisdom, this assumption does not seem to be clear cut... Your suggestion, as I reflect on this, is sound. Made it so.

michaelestrin replied:
`config.ConfigurationStruct` is not an interface.

michaelestrin commented on internal/core/metadata/router.go:
```diff
@@ -38,7 +39,7 @@ func LoadRestRoutes(dic *di.Container) *mux.Router {
 	// Configuration
 	r.HandleFunc(clients.ApiConfigRoute, func(w http.ResponseWriter, r *http.Request) {
-		configHandler(w, container.LoggingClientFrom(dic.Get))
+		configHandler(w, container.LoggingClientFrom(dic.Get), config.ConfigurationStruct{})
```
Why are you passing an empty config.ConfigurationStruct{} here?

Commented here; occurs elsewhere.

akramtexas replied:
I am mortified by this oversight (No excuses here, but bulk refactoring, compounded by inadvertence, contributed to this oversight.) Rectified.

michaelestrin commented on internal/core/metadata/router.go:
```diff
@@ -16,6 +16,7 @@ package metadata
+	"github.com/edgexfoundry/edgex-go/internal/core/metadata/config"
 	metadataContainer "github.com/edgexfoundry/edgex-go/internal/core/metadata/container"
 	"github.com/edgexfoundry/edgex-go/internal/pkg"
 	"github.com/edgexfoundry/edgex-go/internal/pkg/bootstrap/container"
```
For consistency with similar refactorings in other services; alias this as `bootstrapContainer`

michaelestrin commented on internal/core/metadata/router.go:
```diff
@@ -16,6 +16,7 @@ package metadata
 	"net/http"
 
+	"github.com/edgexfoundry/edgex-go/internal/core/metadata/config"
 	metadataContainer "github.com/edgexfoundry/edgex-go/internal/core/metadata/container"
```
For consistency with similar refactorings in other services; this should not be aliased (i.e. it should be internally referenced as `container`)

akramtexas commented:
Recheck

akramtexas commented:
Rebased.

akramtexas commented:
> Blackbox tests fail.
> 
> At least one failure is due to a change from PUT to POST in router.go in changeset [882299a](https://github.com/edgexfoundry/edgex-go/commit/882299a8617a90c44b40bfc847c65706fd16bf4f).
> 
> Unfortunately, you can't currently rely on the blackbox tests in Jenkins for master because they're pulling and running against Fuji artifacts.
> 
> Ask is to work through the blackbox test failures and ensure you understand and fix those failures.
> 
> @akramtexas, @AnthonyMBonafide, @lranjbar

- Updated PR (per `BB testing`) the HTTP method directive in mux router for restUpdateAddressable() and restDeleteProfileByName().
- Rebased.
- Working through the `remaining blackbox test failures` at this time 👍

akramtexas commented:
- Completed working through the remaining blackbox test failures 👍
- Thanks to an invaluable assist by @AnthonyMBonafide, we identified the cause of the remaining blackbox test failures, which were primarily of the following kind, plus one _other_ type which will be mentioned later in this comment. Primarily, those failures concern test artifacts such as the following YAML file:
```
form-data: ENOENT: no such file or directory, open 'sample_profile_duplicated_command.yaml'       
at request                                                                                        
inside "deviceprofile_error_4xx / 142_409 http://localhost:48081/api/v1/deviceprofile/uploadfile" 
```
- That turned out to be a dependence on the directory from where the ("Folder deviceprofile") tests are run from (i.e. the tests, when run from Jenkins, expect for the YAML file above, and three other YAML files to be located at the root of the BB directory.) Once addressed (by updating the _local_ execution context, those failing tests pass).
- Regarding the cause of the remaining blackbox test failures, the one other type (in "Folder devicereport_error_4xx") alluded to earlier has to do with a setup that is expecting from a run of the prior folder of tests (i.e. "Folder deviceprofile"), having to do with this POST:
```
Sub-folder 146 http://localhost:48081/api/v1/deviceprofile/upload
  POST http://localhost:48081/api/v1/deviceprofile/upload [409 Conflict, 203B, 5ms]
```
- Once these testing-setup concerns are addressed, all (core-metadata) BB tests pass, including the following:
```
/usr/local/bin/newman run collections/core-metadata.postman_collection.json -e environment/core-metadata.postman_environment.json --folder deviceprofile -d data/deviceProfileData.json --disable-unicode
newman
core-metadata
Folder deviceprofile
Sub-folder 128 http://localhost:48081/api/v1/deviceprofile/id/{DeviceProfileId}
  GET http://localhost:48081/api/v1/deviceprofile/name/DeviceProfileDeleteByID [200 OK, 650B, 18ms]
  DELETE http://localhost:48081/api/v1/deviceprofile/id/6dbe39f5-a501-4cf0-9c32-9f154d365670 [200 OK, 111B, 8ms]
  Pass  Status code is 200
  Pass  Response time is less than 800ms
  Pass  Has Content-Type
  Pass  Content-Type is application/json
  Pass  Request Objec Deleted
...
...
Sub-folder 146 http://localhost:48081/api/v1/deviceprofile/upload
  POST http://localhost:48081/api/v1/deviceprofile/upload [409 Conflict, 203B, 5ms]
-------------------------------------------------
|                         | executed |   failed |
--------------------------+----------+-----------
|              iterations |        1 |        0 |
--------------------------+----------+-----------
|                requests |       18 |        0 |
--------------------------+----------+-----------
|            test-scripts |       27 |        0 |
--------------------------+----------+-----------
|      prerequest-scripts |       18 |        0 |
--------------------------+----------+-----------
|              assertions |       76 |        0 |
-------------------------------------------------
| total run duration: 826ms                     |
-------------------------------------------------
| total data received: 23.93KB (approx)         |
-------------------------------------------------
| average response time: 4ms                    |
-------------------------------------------------
```
- And these as well:
```
/usr/local/bin/newman run bin/postman-test/collections/core-metadata.postman_collection.json -e bin/postman-test/environment/core-metadata.postman_environment.json --folder devicereport_error_4xx -d bin/postman-test/data/deviceReportData.json --disable-unicode
newman
core-metadata
Folder devicereport_error_4xx
Sub-folder 182 http://localhost:48081/api/v1/devicereport
  POST http://localhost:48081/api/v1/devicereport [400 Bad Request, 162B, 19ms]

Sub-folder 185 http://localhost:48081/api/v1/devicereport
  PUT http://localhost:48081/api/v1/devicereport [400 Bad Request, 162B, 4ms]

Sub-folder 168 http://localhost:48081/api/v1/devicereport/id/{DeviceReportId}
  DELETE http://localhost:48081/api/v1/devicereport/id/57f74b96e4b060e28114165c [404 Not Found, 172B, 3ms]
  Pass  Status code is 404
  Pass  Response time is less than 800ms
  Pass  Has Content-Type
  Pass  Content-Type is text/plain

Sub-folder 171 http://localhost:48081/api/v1/devicereport/name/{DeviceReportName}
  DELETE http://localhost:48081/api/v1/devicereport/name/TestDeviceReportDelByNameNP [404 Not Found, 172B, 3ms]
  Pass  Status code is 404
  Pass  Response time is less than 800ms
  Pass  Has Content-Type
  Pass  Content-Type is text/plain

Sub-folder 188 http://localhost:48081/api/v1/devicereport
  GET http://localhost:48081/api/v1/devicereport [200 OK, 2.57KB, 3ms]

-------------------------------------------------
|                         | executed |   failed |
--------------------------+----------+-----------
|              iterations |        1 |        0 |
--------------------------+----------+-----------
|                requests |        5 |        0 |
--------------------------+----------+-----------
|            test-scripts |        7 |        0 |
--------------------------+----------+-----------
|      prerequest-scripts |        5 |        0 |
--------------------------+----------+-----------
|              assertions |        8 |        0 |
-------------------------------------------------
| total run duration: 268ms                     |
-------------------------------------------------
| total data received: 2.49KB (approx)          |
-------------------------------------------------
| average response time: 6ms                    |
-------------------------------------------------

Process finished with exit code 0
```

akramtexas commented:
- Rebased.
- @michaelestrin this `PR` is ready for re-review 👍

michaelestrin commented on internal/core/metadata/utils.go:
```diff
@@ -15,10 +15,11 @@ package metadata
 
-func checkMaxLimit(limit int) error {
-	if limit > Configuration.Service.MaxResultCount {
+func checkMaxLimit(limit int, config config.ConfigurationStruct) error {
```
Should be passing config by reference instead of value.

Previously called out.  Please ensure all instances have been handled.

akramtexas replied:
Done. Following which, as usual, ran unit tests, ran BB tests 👍

akramtexas commented:
Rebased and ready for review.

michaelestrin commented on cmd/core-metadata/main.go:
```diff
@@ -48,20 +51,28 @@ func main() {
+		},
+		bootstrapContainer.ConfigurationInterfaceName: func(get di.Get) interface{} {
+			return get(container.ConfigurationName)
+		},
```
Please remove this.  See https://github.com/edgexfoundry/edgex-go/issues/2206 merged earlier today.

akramtexas replied:
Done. Rebased and ready for review 👍
